### DISCUSSION TITLE
[DependencyInjection] Find all services that depend on `SomeService`

### DISCUSSION THREAD

**Discussion by ruudk** [symfony/symfony#43787] [Q&A] [answered]
Let's say I have a service called `SomeService`.
I want to get a list of all other services that depend on `SomeService`.
What's the easiest way to do this? 

I hoped to find a method on the `ContainerBuilder` like `getDependedServices(string $id)` but it's not there.

`bin/console debug:container SomeService` doesn't show it. It would be great if it did.

I found this gem in the DependencyInjection but that's not easily reusable:
https://github.com/symfony/symfony/blob/2771d6e65b1a8347f48d3458d98968fb30a73782/src/Symfony/Component/DependencyInjection/Dumper/GraphvizDumper.php#L114-L149

**Comment by stof** [comment, marked as answer]
There is no built-in way to do that in a single call. It requires looping over all services definitions in the ContainerBuilder to find the ones having references to it.

Reusing the ServiceReferenceGraph populated by the AnalyzeServiceReferencesPass might be an option depending on when you need this in the lifetime of the ContainerBuilder, and if it fits the needs (GraphvizDumper does not reuse it for instance).

**Reply by ruudk** [reply]
Thanks @stof, this is exactly what I'm looking for.

I solved my use case with a `CompilerPass` that looks like this:
```php
final class GuardSingleUseGraphQLQueriesPass implements CompilerPassInterface
{
    public function process(ContainerBuilder $container) : void
    {
        $graph = $container->getCompiler()->getServiceReferenceGraph();
        foreach (array_keys($container->findTaggedServiceIds('generated_graphql_client')) as $id) {
            $usages = $graph->getNode($id)->getInEdges();

            if ($usages === []) {
                echo '🤷‍ Operation ' . $id . " is not used!\n";

                continue;
            }

            if (count($usages) === 1) {
                echo '✅ Operation ' . $id . " is used once!\n";

                continue;
            }

            echo sprintf('❌ Operation ' . $id . " is used %d times, it can only be used once:\n", count($usages));
            foreach ($usages as $node) {
                echo sprintf("   > %s\n", $node->getSourceNode()->getId());
            }
        }
    }
}
```

In my bundle I have this:
```php
$container->addCompilerPass(new GuardSingleUseGraphQLQueriesPass(), PassConfig::TYPE_OPTIMIZE);
```

Because the `AnalyzeServiceReferencesPass` already runs before, it populates the `ServiceReferenceGraph` with the data that I need. 😍
